### PR TITLE
JSON test framework to only read JSON files

### PR DESCRIPTION
**Pull request by guscarreon**
If there's a non-json file inside the directory here the JSON test files are run, such a temporary file (vim users can relate), we get an error. This pull request makes the JSON test framework read only files with the `.json` extension.

### Update
The latest commit corrects a bug in the `RunJSONBidderTest(t *testing.T, rootDir string, bidder adapters.Bidder)` that wasn't letting us detect when a non-existent directory was passed as second argument. Such where the cases of Richaudience's `TestGetBuilder(t *testing.T)` and Avocet's `TestJsonSamples(t *testing.T)`.

### CODE REVIEW

File: adapters/adapterstest/test_json.go [hunk 1]
```diff
@@ -65,7 +66,12 @@ func RunJSONBidderTest(t *testing.T, rootDir string, bidder adapters.Bidder) {
 func runTests(t *testing.T, directory string, bidder adapters.Bidder, allowErrors, isAmpTest, isVideoTest bool) {
 	if specFiles, err := os.ReadDir(directory); err == nil {
```
Comment (SyntaxNode) [inline adapters/adapterstest/test_json.go:67]: Did you consider using `filepath.Glob` to use a wildcard pattern in the file listing?

File: adapters/adapterstest/test_json.go [hunk 2]
```diff
@@ -65,7 +66,12 @@ func RunJSONBidderTest(t *testing.T, rootDir string, bidder adapters.Bidder) {
...
 		for _, specFile := range specFiles {
+			if !strings.HasSuffix(specFile.Name(), ".json") {
```
[the remark this answers is not in the document]
Comment (guscarreon) [reply adapters/adapterstest/test_json.go:69]: There's a lot in the `filepath` library. I found the `WalkDir` function particularly useful in order to traverse the directories and reducing our lines of code. Let me know if you agree with its use in the latest delta.

File: adapters/adapterstest/test_json.go (hunk 1 above)
Comment (guscarreon) [reply adapters/adapterstest/test_json.go:67]: We were able to remove this function entirely in the latest delta. Let me know if you agree with the approach.

File: adapters/adapterstest/test_json.go [hunk 3]
```diff
@@ -53,30 +55,44 @@ import (
...
-			specData, err := loadFile(fileName)
+		var approvedPath, allowErrors, isAmpTest, isVideoTest bool
+		switch filepath.Base(filepath.Dir(path)) {
+		case "amp":
+			isAmpTest = true
+			fallthrough
+		case "supplemental":
+			allowErrors = true
+			approvedPath = true
+
```
Comment (hhhjort) [inline adapters/adapterstest/test_json.go:71]: Only supplemental and exemplary lack a `fallthrough`. Why is this?

Comment (hhhjort) [reply adapters/adapterstest/test_json.go:71]: Oh, I think I see the reason now, you are execuding code from multiple cases in one pass.

Comment (guscarreon) [reply adapters/adapterstest/test_json.go:71]: > Oh, I think I see the reason now, you are execuding code from multiple cases in one pass.

Yes but it kind of looks awful. I'm thinking about rolling back to:
```
	switch filepath.Base(filepath.Dir(path)) {
	case "exemplary":
		validDir = true
	case "supplemental":
		validDir = true
		allowErrors = true
	case "amp":
		validDir = true
		allowErrors = true
		isAmpTest = true
	case "video":
		validDir = true
		isVideoTest = true
	case "videosupplemental":
		validDir = true
		allowErrors = true
		isVideoTest = true
	}
```
What do you think

Comment (SyntaxNode) [reply adapters/adapterstest/test_json.go:71]: Agreed that looks awful. Is there to factor out the importance of the directory name? Isn't that just for organization?

Comment (guscarreon) [reply adapters/adapterstest/test_json.go:71]: It's a bit more than only organizational. The test framework in its current state fails a JSON test if said file comes with a `"expectedMakeRequestsErrors"` or a `"expectedMakeBidsErrors"` field but the `allowErrors` flag is set to `false`. Tests in the `supplemental/`, `amp/`, and `videosupplemental/` directories allow errors, but the ones in the `exemplary/` directory, not. I believe in the future we could revisit this.

Comment (guscarreon) [reply adapters/adapterstest/test_json.go:71]: Rolled back to the more verbose but less confusing version of the `switch` statement

File: endpoints/openrtb2/auction_test.go [hunk 4]
```diff
@@ -117,46 +118,64 @@ func TestJsonSampleRequests(t *testing.T) {
...
-						cfg.BlacklistedApps = test.Config.BlacklistedApps
-						cfg.BlacklistedAppMap = test.Config.getBlacklistedAppMap()
-						cfg.BlacklistedAccts = test.Config.BlacklistedAccounts
-						cfg.BlacklistedAcctMap = test.Config.getBlackListedAccountMap()
-						cfg.AccountRequired = test.Config.AccountRequired
-					}
-					cfg.MarshalAccountDefaults()
-					test.endpointType = OPENRTB_ENDPOINT
+			// Test suite will traverse the directory tree recursively and will only consider files with `json` extension
+			if !info.IsDir() && filepath.Ext(info.Name()) == ".json" {
```
Comment (AlexBVolcy) [inline endpoints/openrtb2/auction_test.go:128]: Should `".json"` be extracted to a const value?

File: adapters/adapterstest/test_json.go [hunk 5]
```diff
@@ -53,30 +55,44 @@ import (
...
+			allowErrors = true
+			fallthrough
+		case "video":
+			isVideoTest = true
+			fallthrough
+		case "exemplary":
+			approvedPath = true
+		}
+
+		if !info.IsDir() && approvedPath && filepath.Ext(info.Name()) == ".json" {
```
Comment (AlexBVolcy) [inline adapters/adapterstest/test_json.go:83]: Should `".json"` be extracted to a const variable?

File: adapters/adapterstest/test_json.go [hunk 6]
```diff
@@ -18,6 +20,8 @@ import (
 	"github.com/yudai/gojsondiff/formatter"
 )
 
+const JsonFileExtension string = ".json"
```
Comment (onkarvhanumante) [inline adapters/adapterstest/test_json.go:23]: `JsonFileExtension` can be private 

```
const jsonFileExtension = ".json"
```

Comment (guscarreon) [reply adapters/adapterstest/test_json.go:23]: Good catch. No need to export those. Modified

File: adapters/adapterstest/test_json.go [hunk 7]
```diff
@@ -53,30 +57,45 @@ import (
...
+		case "video":
+			approvedPath = true
+			isVideoTest = true
+		case "videosupplemental":
+			approvedPath = true
+			allowErrors = true
+			isVideoTest = true
+		}
+
+		if !info.IsDir() && approvedPath && filepath.Ext(info.Name()) == JsonFileExtension {
```
[the remark this answers is not in the document]
Comment (guscarreon) [reply adapters/adapterstest/test_json.go]: Looks amazing Onkar. Modified.

**Review by onkarvhanumante — APPROVED** [review]

**Review by AlexBVolcy — APPROVED** [review]

**Pull request merged**